# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/nsHttpConnection.cpp
# Commit: 00bb17e2a6a7
# Full Hash: 00bb17e2a6a73f9c66e861adae426818ff4d8caa
# Author: Kershaw Chang <kershaw@mozilla.com>
# Date: 2023-09-21 21:20:56
# Regressor Bug: 1784261
# File Overlap Count: 1
# Description:
#   Bug 1784261 - Add telemetry to understand the reason of connection closure, r=necko-reviewers,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D188132
# ==============================================================================

diff -r e98350675c5a -r 00bb17e2a6a7 netwerk/protocol/http/nsHttpConnection.cpp
--- a/netwerk/protocol/http/nsHttpConnection.cpp	Thu Sep 21 08:42:26 2023 +0000
+++ b/netwerk/protocol/http/nsHttpConnection.cpp	Thu Sep 21 08:59:03 2023 +0000
@@ -162,6 +162,9 @@
   if (NS_SUCCEEDED(mErrorBeforeConnect)) {
     mSocketTransport->SetEventSink(this, nullptr);
     mSocketTransport->SetSecurityCallbacks(this);
+    ChangeConnectionState(ConnectionState::INITED);
+  } else {
+    SetCloseReason(ToCloseReason(mErrorBeforeConnect));
   }
 
   mTlsHandshaker = new TlsHandshaker(mConnInfo, this);
@@ -493,6 +496,7 @@
         caps));
 
   if (!mExperienced && !trans->IsNullTransaction()) {
+    mHasFirstHttpTransaction = true;
     if (mTlsHandshaker->NPNComplete()) {
       mExperienced = true;
     }
@@ -514,6 +518,12 @@
 
   mTransactionCaps = caps;
   mPriority = pri;
+
+  if (mHasFirstHttpTransaction && mExperienced) {
+    mHasFirstHttpTransaction = false;
+    mExperienceState |= ConnectionExperienceState::Experienced;
+  }
+
   if (mTransaction && (mUsingSpdyVersion != SpdyVersion::NONE)) {
     return AddTransaction(trans, pri);
   }
@@ -556,6 +566,12 @@
   trans->GetSecurityCallbacks(getter_AddRefs(callbacks));
   SetSecurityCallbacks(callbacks);
   mTlsHandshaker->SetupSSL(mInSpdyTunnel, mForcePlainText);
+  if (mTlsHandshaker->NPNComplete()) {
+    // For non-HTTPS connection, change the state to TRANSFERING directly.
+    ChangeConnectionState(ConnectionState::TRANSFERING);
+  } else {
+    ChangeConnectionState(ConnectionState::TLS_HANDSHAKING);
+  }
 
   // take ownership of the transaction
   mTransaction = trans;
@@ -677,8 +693,15 @@
 }
 
 void nsHttpConnection::Close(nsresult reason, bool aIsShutdown) {
-  LOG(("nsHttpConnection::Close [this=%p reason=%" PRIx32 "]\n", this,
-       static_cast<uint32_t>(reason)));
+  LOG(("nsHttpConnection::Close [this=%p reason=%" PRIx32
+       " mExperienceState=%x]\n",
+       this, static_cast<uint32_t>(reason),
+       static_cast<uint32_t>(mExperienceState)));
+
+  if (mConnectionState != ConnectionState::CLOSED) {
+    RecordConnectionCloseTelemetry(reason);
+    ChangeConnectionState(ConnectionState::CLOSED);
+  }
 
   MOZ_ASSERT(OnSocketThread(), "not on socket thread");
   mTlsHandshaker->NotifyClose();
@@ -1213,7 +1236,7 @@
            PR_IntervalToMilliseconds(mTransaction->ResponseTimeout())));
 
       mResponseTimeoutEnabled = false;
-
+      SetCloseReason(ConnectionCloseReason::IDLE_TIMEOUT);
       // This will also close the connection
       CloseTransaction(mTransaction, NS_ERROR_NET_TIMEOUT);
       return UINT32_MAX;
@@ -1238,6 +1261,7 @@
            gHttpHandler->TLSHandshakeTimeout()));
 
       // This will also close the connection
+      SetCloseReason(ConnectionCloseReason::TLS_TIMEOUT);
       CloseTransaction(mTransaction, NS_ERROR_NET_TIMEOUT);
       return UINT32_MAX;
     }
@@ -1562,7 +1586,10 @@
   } else {
     mLastWriteTime = PR_IntervalNow();
     mSocketOutCondition = NS_OK;  // reset condition
-    if (!TunnelSetupInProgress()) mTotalBytesWritten += *countRead;
+    if (!TunnelSetupInProgress()) {
+      mTotalBytesWritten += *countRead;
+      mExperienceState |= ConnectionExperienceState::First_Request_Sent;
+    }
   }
 
   return mSocketOutCondition;
@@ -1748,6 +1775,7 @@
     mSocketInCondition = NS_BASE_STREAM_CLOSED;
   } else {
     mSocketInCondition = NS_OK;  // reset condition
+    mExperienceState |= ConnectionExperienceState::First_Response_Received;
   }
 
   return mSocketInCondition;
@@ -2363,6 +2391,8 @@
     return;
   }
 
+  ChangeConnectionState(ConnectionState::TRANSFERING);
+
   nsCOMPtr<nsITLSSocketControl> tlsSocketControl;
   GetTLSSocketControl(getter_AddRefs(tlsSocketControl));
   if (!tlsSocketControl) {